File: __iOS_BACKUP__/ShapeByte/features/workout/item/domain/ItemSetsHandler.swift
//
//  ItemSetsHandler.swift
//  ShapeByte
//
//  Created by Lang, Stefan [Shape Byte Tech] on 29.07.24.
//

import Foundation
import Combine

class ItemSetsHandler: ObservableObject {
    @Published var state: ItemSetsState = .idle

    private let logger: Logging
    private var currSetHandler: (any ItemSetHandling)?
    private var sets: [ItemSet] = []
    private var statePublisherSink: AnyCancellable?
    private var currSetIndex: Int = -1

    private let timedSetHandler: ItemSetHandling
    private let defaultSetHandler: ItemSetHandling

    init(logger: Logging,
         timedSetHandler: ItemSetHandling,
         defaultSetHandler: ItemSetHandling
    ) {
        self.logger = logger
        self.timedSetHandler = timedSetHandler
        self.defaultSetHandler = defaultSetHandler
    }

    func start(sets: [ItemSet]) {
        if self.sets == sets {
            return
        }

        self.sets = sets
        currSetIndex = -1

        self.state = .started(totalSets: sets.count)
        startNextSet()
    }

    private func startNextSet() {
        let nextSetIndex = currSetIndex + 1

        if nextSetIndex == sets.count {
            finish()
            return
        }

        let nextSet = sets[nextSetIndex]
        currSetIndex = nextSetIndex

        startCoordinationFor(nextSet, onReceivedState: handleSetStateReceived(_:))
    }

    private func handleSetStateReceived(_ state: ItemSet.State) {
        let nextSetIndex = currSetIndex + 1

        switch state {
        case .idle:
            break // TODO: handle
            case .started(let setData):
            logger.logDebug("Started set \(currSetIndex)")
            self.state = ItemSetsState.running(
                currentSet: currSetIndex,
                totalSets: sets.count,
                currentSetProgress: Progress(0),
                totalProgress: totalProgress(currentSetProgress: Progress(0)),
                setData: setData
            )
        case .running(let setData):
            self.state = ItemSetsState.running(
                currentSet: currSetIndex,
                totalSets: sets.count,
                currentSetProgress: setData.progress,
                totalProgress: totalProgress(currentSetProgress: setData.progress),
                setData: setData
            )
        case .paused(let setData):
            self.state = ItemSetsState.paused

        case .finished:
            if nextSetIndex >= sets.count {
                finish()
            } else {
                startNextSet()
            }
        }
    }

    private func totalProgress(currentSetProgress: Progress) -> Progress {
        if currSetIndex < 0 {
            return .zero
        }

        let total = CGFloat((sets.count) * 100)
        let processedSets = CGFloat(currSetIndex) * 100

        let totalProgress = processedSets  + CGFloat(currentSetProgress.absoluteValue)
        let totalProgressRelative = totalProgress / total

        let retVal: Progress = Progress(totalProgressRelative)
        return retVal
    }

    private func finish() {
        self.statePublisherSink?.cancel()
        self.statePublisherSink = nil
        self.currSetHandler = nil

        self.currSetIndex = -1
        self.sets = []

        state = .finished
    }

    private func startCoordinationFor(
        _ itemSet: ItemSet,
        onReceivedState: @escaping (ItemSet.State) -> Void
    ) {
        let nextHandler: any ItemSetHandling

        switch itemSet {
        case .timed:
            nextHandler = timedSetHandler
        default:
            // TODO: log fallback
            nextHandler = defaultSetHandler
        }

        if self.currSetHandler !== nextHandler {
            self.statePublisherSink?.cancel()
            self.statePublisherSink = nil

            self.currSetHandler = nextHandler
            self.statePublisherSink = nextHandler.statePublisher
                .removeDuplicates()
                .sink { state in
                    onReceivedState(state)
            }
        }

        nextHandler.start(set: itemSet)
    }
}

File: shared/src/commonMain/kotlin/de/stefan/lang/shapebyte/features/workout/item/domain/ItemSetsHandler.kt
package de.stefan.lang.shapebyte.features.workout.item.domain

import de.stefan.lang.shapebyte.features.workout.item.data.ItemSet
import de.stefan.lang.shapebyte.features.workout.item.data.ItemSetState
import de.stefan.lang.shapebyte.features.workout.item.domain.timed.TimedItemSetHandler
import de.stefan.lang.shapebyte.utils.Progress
import de.stefan.lang.shapebyte.utils.logging.Loggable
import de.stefan.lang.shapebyte.utils.logging.Logging
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.yield

class ItemSetsHandler(
    override val logger: Logging,
    private val timedSetHandler: TimedItemSetHandler,
    private val defaultSetHandler: DefaultItemSetHandler,
    private val repetitionSetHandler: RepetitionItemSetHandler,
) : Loggable {
    private var sets: List<ItemSet>? = null
    private var currSetHandler: ItemSetHandling? = null
    private var currSetIndex: Int = -1
    private val _stateFlow: MutableStateFlow<ItemSetsState> =
        MutableStateFlow(ItemSetsState.Idle)

    val stateFlow: MutableStateFlow<ItemSetsState> = _stateFlow
    private var stateFlowJob: Job? = null

    fun start(sets: List<ItemSet>, scope: CoroutineScope) {
        if (this.sets == sets) {
            return
        }

        this.sets = sets
        currSetIndex = -1

        _stateFlow.value = ItemSetsState.Started(totalSets = sets.count())
        startNextSet(scope)
    }

    fun setInputValue(value: ItemSetWithInputValue) {
        val inputHandler = currSetHandler as? ItemSetWithInputValueHandling ?: return
        inputHandler.setInputValue(value)
    }

    private fun startNextSet(scope: CoroutineScope) {
        val nextSetIndex = currSetIndex + 1
        val sets = this.sets ?: run {
            logW("(startNextSet) No sets to start")
            emptyList()
        }

        if (nextSetIndex == sets.count()) {
            finish()
            return
        }

        val nextSet = sets[nextSetIndex]
        currSetIndex = nextSetIndex

        startCoordinationFor(nextSet, scope) { data, scope ->
            handleSetStateReceived(data, scope)
        }
    }

    private fun finish() {
        this.currSetHandler = null
        this.currSetIndex = -1
        this.sets = emptyList()

        this._stateFlow.value = ItemSetsState.Finished
    }

    private fun startCoordinationFor(
        itemSet: ItemSet,
        scope: CoroutineScope,
        onReceivedState: suspend (ItemSetState, CoroutineScope) -> Unit,
    ) {
        val nextHandler: ItemSetHandling = when (itemSet) {
            is ItemSet.Timed -> timedSetHandler
            is ItemSet.Repetition -> repetitionSetHandler
            else -> defaultSetHandler
        }

        if (currSetHandler !== nextHandler) {
            stateFlowJob?.cancel()
            stateFlowJob = null
            currSetHandler = nextHandler

            stateFlowJob = scope.launch {
                nextHandler.stateFlow.collect { state ->
                    onReceivedState(state, this)

                    if (!coroutineContext.isActive || stateFlow.value is ItemSetsState.Finished) {
                        cancel()
                    }

                    yield()
                }
            }
        }

        nextHandler.start(set = itemSet, scope)
    }

    private fun handleSetStateReceived(state: ItemSetState, scope: CoroutineScope) {
        val nextSetIndex = currSetIndex + 1
        val sets = this.sets ?: return

        when (state) {
            is ItemSetState.Idle -> {}

            is ItemSetState.Started -> {
                logD("Started set $currSetIndex")
                _stateFlow.value = ItemSetsState.Running.SetStarted(
                    currentSetIndex = currSetIndex,
                    totalSets = sets.count(),
                    setData = state.setData,
                )
            }

            is ItemSetState.Running -> {
                logD("Running set $currSetIndex")
                _stateFlow.value = ItemSetsState.Running.SetRunning(
                    currentSetIndex = currSetIndex,
                    totalSets = sets.count(),
                    currentSetProgress = state.setData.progress,
                    totalProgress = totalProgress(currentSetProgress = state.setData.progress),
                    setData = state.setData,
                )
            }

            is ItemSetState.Paused -> {
                logD("Pausing set $currSetIndex")
                _stateFlow.value = ItemSetsState.Paused
            }

            is ItemSetState.Finished -> {
                logD("Finished set $currSetIndex")
                _stateFlow.value = ItemSetsState.Running.SetFinished(
                    currentSetIndex = currSetIndex,
                    totalSets = sets.count(),
                    setData = state.setData,
                )

                if (nextSetIndex >= sets.count()) {
                    finish()
                } else {
                    startNextSet(scope)
                }
            }
        }
    }

    private fun totalProgress(currentSetProgress: Progress): Progress {
        if (currSetIndex < 0) {
            return Progress.ZERO
        }

        val sets = this.sets ?: return Progress.ZERO
        val total = sets.count() * Progress.ABSOLUTE
        val processedSets = currSetIndex * Progress.ABSOLUTE

        val totalProgress = processedSets + currentSetProgress.absoluteValue
        val totalProgressRelative = (totalProgress.toFloat() / total.toFloat())

        val retVal = Progress(totalProgressRelative)
        return retVal
    }
}
